File: skills/self/self-improve/lib/harness_health.py
#!/usr/bin/env python3
"""harness_health.py -- Pure: tool-use health aggregation over already-parsed ledger records.

anicca-harness-tooluse-health R8: Python mirror of runtime/loop/harness-health.mjs's R1-R5
(classify_layer, compute_slot_health, compute_brain_transport_health, compute_harness_health,
should_escalate). Identical contract to the JS twin, proven via a SHARED fixture
(runtime/loop/__tests__/fixtures/harness-health-parity.json), not shared code. Reuses
ledger_metrics.load_ledger_rows for jsonl parsing (never a new/duplicated jsonl loader).

INV-NO-JUDGMENT: pure bookkeeping over the already-deterministic `kind` enum index.mjs itself
assigns -- never a new judgment. INV-LOOPDETECT-SEPARATE: `loop_detect` is excluded from
compute_slot_health's subset by a `kind`-membership check, regardless of its own `slot` field.

See specs/behavioral-spec.md and specs/verification-architecture.md
(anicca-project/.vcsdd/features/anicca-harness-tooluse-health/) for the full requirement text.
"""
import os
from datetime import datetime, timezone

# R5's documented default: same `Number(process.env.X) || N` fallback idiom already used by
# catalog-gate.mjs::DEFAULT_BOOTSTRAP_RESERVE_USDC (and its own JS twin, harness-health.mjs).
try:
    DEFAULT_STREAK_THRESHOLD = int(os.environ.get("HARNESS_HEALTH_STREAK_THRESHOLD") or 0) or 5
except (TypeError, ValueError):
    DEFAULT_STREAK_THRESHOLD = 5

CLEAN_KINDS = {"wake", "narrate", "shutdown"}
# R2's slot-health subset: `loop_detect` is deliberately NOT a member even though its records carry
# a `slot` field -- INV-LOOPDETECT-SEPARATE.
SLOT_HEALTH_KINDS = {"wake", "skill_missing", "skill_timeout", "skill_error"}
# R3's brain-transport streak reset set -- narrower than CLEAN_KINDS (shutdown never resets it).
BRAIN_TRANSPORT_RESET_KINDS = {"wake", "narrate"}


def classify_layer(record):
    """R1 -- maps record['kind'] to exactly one health layer. Fail-safe: any unrecognized kind
    (including loop_detect, deliberately out of this mapping per INV-LOOPDETECT-SEPARATE) maps to
    'unknown', never raises."""
    kind = record.get("kind") if isinstance(record, dict) else None
    if kind in CLEAN_KINDS:
        return "clean"
    if kind == "wake_error":
        return "brain_transport"
    if kind == "skill_missing":
        return "tool_missing"
    if kind == "skill_timeout":
        return "tool_timeout"
    if kind == "skill_error":
        return "tool_logic"
    return "unknown"


def compute_slot_health(records, slot):
    """R2 -- per-slot tool-call health over the subset of `records` carrying that exact `slot`
    value AND kind in SLOT_HEALTH_KINDS (loop_detect excluded regardless of its own slot field).
    `records` is assumed chronologically ordered. Returns None if `slot` is absent entirely."""
    arr = records if isinstance(records, list) else []
    subset = [r for r in arr if isinstance(r, dict) and r.get("slot") == slot and r.get("kind") in SLOT_HEALTH_KINDS]
    if not subset:
        return None

    wakes = len(subset)
    failures = 0
    last_failure_layer = None
    last_failure_ts = None
    last_failure_wake_id = None

    for r in subset:
        layer = classify_layer(r)
        if layer != "clean":
            failures += 1
            last_failure_layer = layer
            last_failure_ts = r.get("ts")
            last_failure_wake_id = r.get("wake_id")

    consecutive_failure_streak = 0
    for r in reversed(subset):
        if classify_layer(r) == "clean":
            break
        consecutive_failure_streak += 1

    return {
        "slot": slot,
        "wakes": wakes,
        "failures": failures,
        "failureRate": failures / wakes,
        "consecutiveFailureStreak": consecutive_failure_streak,
        "lastFailureLayer": last_failure_layer,
        "lastFailureTs": last_failure_ts,
        "lastFailureWakeId": last_failure_wake_id,
    }


def compute_brain_transport_health(records):
    """R3 -- loop-level health of ALL wake_error records across the WHOLE ledger, never conflated
    with any single slot's health. Streak scans the WHOLE records list (reset by wake/narrate only);
    any other kind is skipped (neither counts nor resets)."""
    arr = records if isinstance(records, list) else []
    wake_errors = [r for r in arr if isinstance(r, dict) and r.get("kind") == "wake_error"]
    failures = len(wake_errors)

    consecutive_failure_streak = 0
    for r in reversed(arr):
        if not isinstance(r, dict):
            continue
        kind = r.get("kind")
        if kind == "wake_error":
            consecutive_failure_streak += 1
            continue
        if kind in BRAIN_TRANSPORT_RESET_KINDS:
            break
        # any other kind: skip, does not count and does not reset

    last = wake_errors[-1] if wake_errors else None
    return {
        "failures": failures,
        "consecutiveFailureStreak": consecutive_failure_streak,
        "lastFailureTs": last.get("ts") if last else None,
        "lastFailureWakeId": last.get("wake_id") if last else None,
    }


def should_escalate(consecutive_failure_streak, threshold):
    """R5 -- pure boundary predicate, never invokes any repair action itself (R10)."""
    return float(consecutive_failure_streak) >= float(threshold)


def compute_harness_health(records, streak_threshold=None, now=None):
    """R4 -- whole-ledger report shape covering every distinct slot value present in the R2 subset
    (SLOT_HEALTH_KINDS), plus loop-level brain-transport health, each entry annotated with R5's
    `escalate`. Never raises, even for empty/absent `records`.

    `now` is an optional datetime override for generatedAt's wall-clock stamp (test-injection idiom
    mirroring the JS twin's `now` option); defaults to the current UTC instant when omitted."""
    arr = records if isinstance(records, list) else []
    threshold = streak_threshold if streak_threshold is not None else DEFAULT_STREAK_THRESHOLD

    slots = []
    for r in arr:
        if isinstance(r, dict) and r.get("kind") in SLOT_HEALTH_KINDS and r.get("slot") is not None:
            if r["slot"] not in slots:
                slots.append(r["slot"])

    per_slot = {}
    for slot in slots:
        health = compute_slot_health(arr, slot)
        per_slot[slot] = {**health, "escalate": should_escalate(health["consecutiveFailureStreak"], threshold)}

    brain_health = compute_brain_transport_health(arr)
    brain_transport = {**brain_health, "escalate": should_escalate(brain_health["consecutiveFailureStreak"], threshold)}

    generated_at = now if now is not None else datetime.now(timezone.utc)
    return {
        "perSlot": per_slot,
        "brainTransport": brain_transport,
        "generatedAt": generated_at.isoformat().replace("+00:00", "Z"),
    }

File: runtime/loop/harness-health.mjs
/**
 * harness-health.mjs — Pure: tool-use health aggregation over already-parsed ledger records.
 *
 * anicca-harness-tooluse-health R1-R5: classifyLayer, computeSlotHealth,
 * computeBrainTransportHealth, computeHarnessHealth, shouldEscalate. Every function here takes
 * already-parsed ledger record arrays and returns plain objects — none opens a file, spawns a
 * process, or reads process.env except DEFAULT_STREAK_THRESHOLD's one module-level constant (same
 * `Number(process.env.X) || N` idiom already used by catalog-gate.mjs::DEFAULT_BOOTSTRAP_RESERVE_USDC).
 *
 * INV-NO-JUDGMENT: this is pure bookkeeping over the ALREADY-deterministic `kind` enum index.mjs
 * itself assigns via if-else on exitCode/timedOut/notFound — never a new judgment.
 * INV-LOOPDETECT-SEPARATE: `loop_detect` already has its own consecutive-streak + exponential
 * cooldown (index.mjs lines 180-284); it is explicitly excluded from computeSlotHealth's subset by a
 * `kind`-membership check, regardless of whether its (shared) `slot` field matches.
 *
 * See specs/behavioral-spec.md and specs/verification-architecture.md
 * (anicca-project/.vcsdd/features/anicca-harness-tooluse-health/) for the full requirement text.
 */

// REQ-R5's documented default: same `Number(process.env.X) || N` fallback idiom already used for
// DEFAULT_BOOTSTRAP_RESERVE_USDC in catalog-gate.mjs.
export const DEFAULT_STREAK_THRESHOLD = Number(process.env.HARNESS_HEALTH_STREAK_THRESHOLD) || 5;

const CLEAN_KINDS = new Set(['wake', 'narrate', 'shutdown']);
// R2's slot-health subset: `loop_detect` is deliberately NOT a member of this set even though its
// records carry a `slot` field (index.mjs:281) — INV-LOOPDETECT-SEPARATE.
const SLOT_HEALTH_KINDS = new Set(['wake', 'skill_missing', 'skill_timeout', 'skill_error']);
// R3's brain-transport streak reset set — narrower than CLEAN_KINDS (shutdown never resets it,
// consistent with behavioral-spec.md R3's literal text: "reset by any record whose kind ∈ {wake, narrate}").
const BRAIN_TRANSPORT_RESET_KINDS = new Set(['wake', 'narrate']);

/**
 * R1 — classifyLayer(record): maps `record.kind` to exactly one health layer. Fail-safe: any
 * unrecognized kind (including `loop_detect`, which is deliberately out of this mapping per
 * INV-LOOPDETECT-SEPARATE) maps to 'unknown', never throws.
 *
 * @param {{kind?: string}} record
 * @returns {'clean'|'brain_transport'|'tool_missing'|'tool_timeout'|'tool_logic'|'unknown'}
 */
export function classifyLayer(record) {
  const kind = record && record.kind;
  if (CLEAN_KINDS.has(kind)) return 'clean';
  if (kind === 'wake_error') return 'brain_transport';
  if (kind === 'skill_missing') return 'tool_missing';
  if (kind === 'skill_timeout') return 'tool_timeout';
  if (kind === 'skill_error') return 'tool_logic';
  return 'unknown';
}

/**
 * R2 — computeSlotHealth(records, slot): per-slot tool-call health over the subset of `records`
 * carrying that exact `slot` value AND `kind` ∈ SLOT_HEALTH_KINDS (loop_detect excluded regardless
 * of its own `slot` field — INV-LOOPDETECT-SEPARATE). `records` is assumed chronologically ordered
 * (the same order `readLedgerLines`/`recentLedgerLines` already produce).
 *
 * @param {object[]} records
 * @param {string} slot
 * @returns {{slot:string,wakes:number,failures:number,failureRate:number,consecutiveFailureStreak:number,lastFailureLayer:string|null,lastFailureTs:number|null,lastFailureWakeId:string|null}|null}
 */
export function computeSlotHealth(records, slot) {
  const arr = Array.isArray(records) ? records : [];
  const subset = arr.filter((r) => r && r.slot === slot && SLOT_HEALTH_KINDS.has(r.kind));
  if (subset.length === 0) return null;

  const wakes = subset.length;
  let failures = 0;
  let lastFailureLayer = null;
  let lastFailureTs = null;
  let lastFailureWakeId = null;

  for (const r of subset) {
    const layer = classifyLayer(r);
    if (layer !== 'clean') {
      failures += 1;
      lastFailureLayer = layer;
      lastFailureTs = r.ts != null ? r.ts : null;
      lastFailureWakeId = r.wake_id != null ? r.wake_id : null;
    }
  }

  let consecutiveFailureStreak = 0;
  for (let i = subset.length - 1; i >= 0; i--) {
    if (classifyLayer(subset[i]) === 'clean') break;
    consecutiveFailureStreak += 1;
  }

  return {
    slot,
    wakes,
    failures,
    failureRate: failures / wakes,
    consecutiveFailureStreak,
    lastFailureLayer,
    lastFailureTs,
    lastFailureWakeId,
  };
}

/**
 * R3 — computeBrainTransportHealth(records): loop-level health of ALL `wake_error` records (which
 * never carry `slot` — R1) across the WHOLE ledger, never conflated with any single slot's health.
 * The trailing streak is scanned over the WHOLE `records` array (not just the wake_error subset)
 * because its reset condition depends on OTHER kinds (`wake`/`narrate`) that only exist outside that
 * subset; any other kind (skill_error, skill_missing, skill_timeout, loop_detect, shutdown) is
 * skipped while scanning — it neither counts nor resets the streak.
 *
 * @param {object[]} records
 * @returns {{failures:number,consecutiveFailureStreak:number,lastFailureTs:number|null,lastFailureWakeId:string|null}}
 */
export function computeBrainTransportHealth(records) {
  const arr = Array.isArray(records) ? records : [];
  const wakeErrors = arr.filter((r) => r && r.kind === 'wake_error');
  const failures = wakeErrors.length;

  let consecutiveFailureStreak = 0;
  for (let i = arr.length - 1; i >= 0; i--) {
    const r = arr[i];
    if (!r) continue;
    if (r.kind === 'wake_error') {
      consecutiveFailureStreak += 1;
      continue;
    }
    if (BRAIN_TRANSPORT_RESET_KINDS.has(r.kind)) break;
    // any other kind: skip, does not count and does not reset
  }

  const last = wakeErrors[wakeErrors.length - 1];
  return {
    failures,
    consecutiveFailureStreak,
    lastFailureTs: last ? (last.ts != null ? last.ts : null) : null,
    lastFailureWakeId: last ? (last.wake_id != null ? last.wake_id : null) : null,
  };
}

/**
 * R5 — shouldEscalate(consecutiveFailureStreak, threshold): pure boundary predicate, never invokes
 * any repair action itself (R10).
 *
 * @param {number} consecutiveFailureStreak
 * @param {number} threshold
 * @returns {boolean}
 */
export function shouldEscalate(consecutiveFailureStreak, threshold) {
  return Number(consecutiveFailureStreak) >= Number(threshold);
}

/**
 * R4 — computeHarnessHealth(records, opts): whole-ledger report shape covering every distinct slot
 * value present in the R2 subset (SLOT_HEALTH_KINDS), plus loop-level brain-transport health, each
 * entry annotated with R5's `escalate`. Never throws, even for empty/absent `records`.
 *
 * `now` is an optional epoch-ms (or Date) override for `generatedAt`'s wall-clock stamp — the same
 * test-injection idiom this codebase already uses elsewhere (ANICCA_BALANCE_OVERRIDE, CLAUDE_BIN);
 * defaults to Date.now() when omitted. Not part of the R1-R5 deterministic contract itself.
 *
 * @param {object[]} records
 * @param {{streakThreshold?: number, now?: number|Date}} [opts]
 * @returns {{perSlot: Record<string, object>, brainTransport: object, generatedAt: string}}
 */
export function computeHarnessHealth(records, opts = {}) {
  const arr = Array.isArray(records) ? records : [];
  const { streakThreshold, now } = opts || {};
  const threshold = streakThreshold != null ? streakThreshold : DEFAULT_STREAK_THRESHOLD;

  const slots = [];
  for (const r of arr) {
    if (r && SLOT_HEALTH_KINDS.has(r.kind) && r.slot != null && !slots.includes(r.slot)) {
      slots.push(r.slot);
    }
  }

  const perSlot = {};
  for (const slot of slots) {
    const health = computeSlotHealth(arr, slot);
    perSlot[slot] = { ...health, escalate: shouldEscalate(health.consecutiveFailureStreak, threshold) };
  }

  const brainHealth = computeBrainTransportHealth(arr);
  const brainTransport = { ...brainHealth, escalate: shouldEscalate(brainHealth.consecutiveFailureStreak, threshold) };

  return {
    perSlot,
    brainTransport,
    generatedAt: new Date(now != null ? now : Date.now()).toISOString(),
  };
}

/**
 * R6 helper — capFailureDetail(text): whitespace-collapse + cap at 4000 chars, the SAME shared pure
 * step index.mjs::runOneWake applies identically to BOTH the tool_missing/tool_timeout/tool_logic
 * branch (already-redacted skillResult.output) and the brain_transport branch (its own
 * redactPrivateKeyPatterns(err.message) output) before appending to harness-failures.jsonl. Distinct
 * from — and never applied to — the existing 900-char `result` field on ledger.jsonl
 * (INV-NO-PROMPT-REGRESSION).
 *
 * @param {unknown} text
 * @returns {string}
 */
export function capFailureDetail(text) {
  const s = typeof text === 'string' ? text : String(text == null ? '' : text);
  return s.replace(/\s+/g, ' ').slice(0, 4000);
}
